build/room_ensurer/room_ensurer.py

```diff
@@ -0,0 +1,300 @@
+    def __init__(
+        self,
+        username: str,
+        password: str,
+        own_server_name: str,
+        known_servers_url: Optional[str] = None,
+    ):
+        self._username = username
+        self._password = password
+        self._own_server_name = own_server_name
+
+        if known_servers_url is None:
+            known_servers_url = DEFAULT_MATRIX_KNOWN_SERVERS[Environment.PRODUCTION]
+
+        self._known_servers: Dict[str, str] = {
+            urlparse(server_url).netloc: server_url
+            for server_url in sorted(get_matrix_servers(known_servers_url))
+        }
+        if not self._known_servers:
+            raise RuntimeError(f"No known servers found from list at {known_servers_url}.")
+        self._first_server_name = list(self._known_servers.keys())[0]
+        self._is_first_server = own_server_name == self._first_server_name
+        self._apis: Dict[str, GMatrixHttpApi] = self._connect_all()
+        self._own_api = self._apis[own_server_name]
+        log.debug(
+            "Room ensurer initialized",
+            own_server_name=own_server_name,
+            known_servers=self._known_servers.keys(),
+            first_server_name=self._first_server_name,
+            is_first_server=self._is_first_server,
+        )
+
+    def ensure_rooms(self) -> None:
+        exceptions = {}
+        for network in Networks:
+            for alias_fragment in [
+                DISCOVERY_DEFAULT_ROOM,
+                MONITORING_BROADCASTING_ROOM,
+                PATH_FINDING_BROADCASTING_ROOM,
+            ]:
+                try:
+                    self._ensure_room_for_network(network, alias_fragment)
+                except (MatrixError, EnsurerError) as ex:
+                    log.exception(f"Error while ensuring room for {network.name}.")
+                    exceptions[network] = ex
+        if exceptions:
+            log.error("Exceptions happened", exceptions=exceptions)
+
+    def _ensure_room_for_network(self, network: Networks, alias_fragment: str) -> None:
+        log.info(f"Ensuring {alias_fragment} room for {network.name}")
+        room_alias_prefix = make_room_alias(ChainID(network.value), alias_fragment)
+        room_infos: Dict[str, Optional[RoomInfo]] = {
+            server_name: self._get_room(server_name, room_alias_prefix)
+            for server_name in self._known_servers.keys()
+        }
+        first_server_room_info = room_infos[self._first_server_name]
+
+        if not first_server_room_info:
+            log.warning("First server room missing")
+            if self._is_first_server:
+                log.info("Creating room", server_name=self._own_server_name)
+                first_server_room_info = self._create_room(
+                    self._own_server_name, room_alias_prefix
+                )
+                room_infos[self._first_server_name] = first_server_room_info
+            else:
+                raise EnsurerError("First server room missing.")
```

Comment:
What happens in this case?

Reply:
The exception gets caught in `main()` below and after a 60s sleep the process is retried.

Reply:
But is it created at some point? Or will this just end in a loop?

Reply:
Yes, see line 133 above.
Since the script is supposed to run on all servers eventually the room schools appear.